modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuildController.java

```diff
@@ -0,0 +1,299 @@
+            if (exception != null) {
+                return failedFuture(exception);
+            }
+
+            return onIndexDrop(((DropIndexEventParameters) parameters)).thenApply(unused -> false);
+        });
+
+        placementDriver.listen(PrimaryReplicaEvent.PRIMARY_REPLICA_ELECTED, (parameters, exception) -> {
+            if (exception != null) {
+                return failedFuture(exception);
+            }
+
+            return onPrimaryReplicaElected(parameters).thenApply(unused -> false);
+        });
+    }
+
+    private CompletableFuture<?> onIndexCreate(CreateIndexEventParameters parameters) {
+        return inBusyLockAsync(busyLock, () -> {
+            var startBuildIndexFutures = new ArrayList<CompletableFuture<?>>();
+
+            for (TablePartitionId primaryReplicaId : primaryReplicaIds) {
+                if (primaryReplicaId.tableId() == parameters.indexDescriptor().tableId()) {
+                    CompletableFuture<?> startBuildIndexFuture = getMvTableStorageFuture(parameters.causalityToken(), primaryReplicaId)
+                            .thenCompose(mvTableStorage -> awaitPrimaryReplicaFowNow(primaryReplicaId)
+                                    .thenAccept(replicaMeta -> tryStartBuildIndex(
+                                            primaryReplicaId,
+                                            parameters.indexDescriptor(),
+                                            mvTableStorage,
+                                            replicaMeta
+                                    ))
+                            );
+
+                    startBuildIndexFutures.add(startBuildIndexFuture);
+                }
+            }
+
+            return allOf(startBuildIndexFutures.toArray(CompletableFuture[]::new));
+        });
+    }
+
+    private CompletableFuture<?> onIndexDrop(DropIndexEventParameters parameters) {
+        return inBusyLockAsync(busyLock, () -> {
+            indexBuilder.stopBuildingIndexes(parameters.indexId());
+
+            return completedFuture(null);
+        });
+    }
+
+    private CompletableFuture<?> onPrimaryReplicaElected(PrimaryReplicaEventParameters parameters) {
+        return inBusyLockAsync(busyLock, () -> {
+            TablePartitionId primaryReplicaId = (TablePartitionId) parameters.groupId();
+
+            if (isLocalNode(parameters.leaseholder())) {
+                primaryReplicaIds.add(primaryReplicaId);
+
+                // It is safe to get the latest version of the catalog because the PRIMARY_REPLICA_ELECTED event is handled on the
+                // metastore thread.
+                int catalogVersion = catalogService.latestCatalogVersion();
+
+                return getMvTableStorageFuture(parameters.causalityToken(), primaryReplicaId)
+                        .thenCompose(mvTableStorage -> awaitPrimaryReplicaFowNow(primaryReplicaId)
+                                .thenAccept(replicaMeta -> tryStartBuildIndexesForNewPrimaryReplica(
```

Comment:
What future are you talking about?
In general, this is not so important; simple operations are performed here; the physical building of indexes is performed in the `org.apache.ignite.internal.table.distributed.index.IndexBuilder#executor`.

Reply:
I'm talking about "awaitPrimaryReplicaFowNow", I thought you can tell that from the context.
What kind of simple operations are you talking about? Scanning through partitions and reading last built row from them? This is IO, it may not be as fast as you hope. Or you mean other operations?

Reply:
Please look at implementation `org.apache.ignite.internal.table.distributed.index.IndexBuilder#startBuildIndex`, essentially we are just putting the task of building the index in a separate thread pool. I can point this out in documentation of `IndexBuilder#startBuildIndex`, or what do you suggest?

Reply:
Ok, I was confused by what "startBuild" does. I guess I should rely more on the documentation of methods, rather than their names

Reply:
You can rename it to `scheduleBuildIndex` (maybe sounds like something to plan for later) or `addBuildIndexTask`, what do you think?

Reply:
fix it